### ISSUE TITLE
RFE: anon followers

### ISSUE THREAD

**Issue by nirik** [repoSpanner/repoSpanner#33]
It might be nice to have a way for anon people to just connect to the repospanner cluster to copy the repos, but never write to them. This would allow people to mirror all or a subset of repos they are interested in and be more efficent than grokmirror or just pulling in a bunch of repos.

**Comment by puiterwijk** [comment]
Given that this is pretty specific to environments (in general, you can't say that all repositories on a cluster have the same user base, we might get the same in Fedora with `ansible` vs `ansible-private`, and would suggest to do this based on post-receive hooks, that send messages to a message bus for example.
We could add e.g. add a websocket for public repositories, but given that at least Fedora already has a message bus, I'm not sure we really need this at this moment.

**Comment by igor-raits** [comment]
@puiterwijk so how would an average user make a mirror of all Fedora repos? And also update them (without re-cloning all of them)?

**Comment by smooge** [comment]
I don't think repospanner is meant to be a service which is supposed to allow this. It is like asking, why doesn't github allow anyone connect to their backend storage tools and get what they want versus using the tool github is expecting people to use. [Or allowing anyone to connect to Fedora's NFS servers in a read-only method just so they can get lookaside cache.] It is using a tool which is meant for a specific purpose in a way that it isn't and isn't designed to be used.

**Comment by igor-raits** [comment]
Well, will there be some easy way to deploy some service in Fedora infra based on repoSpanner which would allow doing mirrors?

It is kinda complicated right now, because grok-mirror is half-broken most of the time.. And since we are moving to repoSpanner, it would make sense to replace it by something what works.

**Comment by puiterwijk** [comment]
@ignatenkobrain As an average user, you could do what you can do right now:

1. Clone all repositories once (either just cloning them all, or using the git seed)
2. Run a fedmsg listener that listens for Git commits/pushes, and do a git pull on the specific repository the message was sent for.

This means you would not be tied in to anything repoSpanner-specific.

**Comment by igor-raits** [comment]
> Run a fedmsg listener that listens for Git commits/pushes, and do a git pull on the specific repository the message was sent for.

This assumes that I need my laptop running and being connected to the internet all the time… Which is not something what I would like to have.

**Comment by puiterwijk** [comment]
> This assumes that I need my laptop running and being connected to the internet all the time… Which is not something what I would like to have.

Nah, fedmsg-hub will record the last message it processed, and on a start retrieve all mesages sent since then from datagrepper and process those.

**Comment by smooge** [comment]
Again this is looking at a tool which does one thing and then saying 'well I want it to do this other thing too'. And then adding another thing and another thing and.. and in in 3-4 months we are back to seeing everything broken and overly complicated again. repoSpanner is meant to duplicate repositories for the backend of specific front end tools. If it doesn't fit your needs then we can look at building a separate tool which does but adding this in requires making changes in a lot of places.

**Comment by igor-raits** [comment]
> If it doesn't fit your needs then we can look at building a separate tool which does but adding this in requires making changes in a lot of places.

Note that I'm not person opened this issue, I just spoke to @nirik over the DevConf.cz about issues with grok-mirror and he said that we probably want to handle this by repoSpanner, so this is how we are here.

I'm fine with any of the tools as long as they are performing their task well.
